### PR TITLE
Focus-chain update

### PR DESCRIPTION
**Pull request by xarvic**
This is a fix of the part of #1716 @cmyr and i agreed on:

- `should_propagate_to_hidden()` now returns `false` for `Lifecycle::BuildFocusChain`
- `Lifecycle::BuildFocusChain` requests a a resign of focus if `state.has_focus` is set but neither one of its hidden children nor it self has focus.

The other point: What should happen if try to focus a WidgetID which is not allowed to recieve focus like `ctx.set_focus(WidgetID::new())` or focusing a hidden widget is not addressed here.

### CODE REVIEW

File: druid/src/core.rs [hunk 1]
```diff
@@ -1082,6 +1088,17 @@ impl<T: Data, W: Widget<T>> WidgetPod<T, W> {
             // Update focus-chain of our parent
             LifeCycle::BuildFocusChain => {
                 self.state.update_focus_chain = false;
+
+                // had_focus is the old focus value. state.has_focus was repaced with ctx.is_focused().
+                // Therefore is had_focus is true but state.has_focus is false the widgets, which is
+                // currently focused is not part of the functional tree anymore
+                // (Lifecycle::BuildFocusChain.should_propagate_to_hidden() is false!) and should
+                // resign the focus.
```
Comment (cmyr) [inline druid/src/core.rs:1096]:
```suggestion
                // had_focus is the old focus value. state.has_focus was repaced with ctx.is_focused().
                // Therefore if had_focus is true but state.has_focus is false then the widget which is
                // currently focused is not part of the functional tree anymore
                // (Lifecycle::BuildFocusChain.should_propagate_to_hidden() is false!) and should
                // resign the focus.
```

File: druid/src/core.rs [hunk 2]
```diff
@@ -1082,6 +1088,17 @@ impl<T: Data, W: Widget<T>> WidgetPod<T, W> {
...
+                if had_focus && !self.state.has_focus {
```
Comment (cmyr) [inline druid/src/core.rs:1097]: I'm confused; we haven't done `merge_up` yet, so how would `has_focus` have changed?

**Review by cmyr — COMMENTED** [review]
I'm a bit confused here, and will need to take a second look (or maybe we should have a quick chat about it) I'm having a slow time remembering the context for this. In any case, one question inline.

File: druid/src/core.rs (hunk 2 above)
Comment (xarvic) [reply druid/src/core.rs:1097]: The `merge_up` of our children happened already in line 1058 / 1062. If someone hides a focused widget and then calls `children_changed()`, we clear `has_focus` in `BuildFocusChain` if has_clear is not set after the calls in 1058 and 1062 we know, that the focused widget is not reachable

**Review by cmyr — APPROVED** [review]
Okay, that explanation works for me, thanks!

**Pull request merged**